### ISSUE TITLE
make_number_formatter() causing UnicodeDecodeError on FreeBSD

### ISSUE THREAD

**Issue by Kirkman** [wireservice/agate#717]
I have some code that I run under Python 2.7.15 in two different environments: Mac OS X and FreeBSD.

(I'm using Agate 1.6.1 and csvkit 1.0.3 in both environments)

The trouble is using `.print_table()` on columns with long decimals. It's working fine on my Mac, but when I run the same code under FreeBSD, I get errors like this:

```
  File "csv_to_json.py", line 89, in read_incident_csv
    incident_table.print_table( max_columns=15, max_column_width=6, output=output_device )
  File "/home/newsroom/.virtualenvs/crime/lib/python2.7/site-packages/agate/table/print_table.py", line 156, in print_table
    write_row(formatted_row)
  File "/home/newsroom/.virtualenvs/crime/lib/python2.7/site-packages/agate/table/print_table.py", line 140, in write_row
    write('%s%s%s' % (v_line, text, v_line))
  File "/home/newsroom/.virtualenvs/crime/lib/python2.7/site-packages/agate/table/print_table.py", line 121, in write
    output.write(line + '\n')
UnicodeEncodeError: 'ascii' codec can't encode character u'\u2026' in position 67: ordinal not in range(128)
```

The issue seems to be an ellipsis character which is added to my lat/lng columns (which are very long decimals) by [this line](https://github.com/wireservice/agate/blob/master/agate/utils.py#L169) in `make_number_formatter()`:

```
    ellipsis = u'…' if add_ellipsis else u''
```

If I edit `utils.py` and change that Unicode ellipsis to be some other simple ASCII character, like an exclamation mark, then `print_table()` works fine in my FreeBSD environment.

I don't really understand why this is going wrong. Maybe there's something I could do in my environment to avoid it? FWIW, I checked the locales in each environment, and FreeBSD reports the same values as my Mac:

```
LANG=
LC_CTYPE="C"
LC_COLLATE="C"
LC_TIME="C"
LC_NUMERIC="C"
LC_MONETARY="C"
LC_MESSAGES="C"
LC_ALL=
```

On a whim, I tried passing a locale value of `en-US.UTF-8` to `print_table()`, but it didn't make any difference.

However, I want to point out that it seems weird to have this hard-coded ellipsis character in the `make_number_formatter()` function, particularly when you [provide a place to specify custom ellipsis characters](https://github.com/wireservice/agate/blob/master/agate/config.py#L56) in `config.py`. Shouldn't `make_number_formatter()` use that value instead?

**Comment by Kirkman** [comment]
Just want to update this 1.5 years later. I am now running into the same problem with the  `print_table()` ellipsis character under a new environment.

We are now using CentOS Linux, and I still have many scripts that use Agate. In those scripts, I generate an `output_device` variable like this, in order to suppress Agate's `print_table()` when a script is running from a cronjob:
```
if sys.stdout.isatty():
    output_device = sys.stdout
else:
    output_device = open(os.devnull, 'w')
```

Then I can do:
```
all_tables.print_table( max_columns=10, max_column_width=6, output=output_device )
```

Anyway, my scripts run fine when I run them myself from the command line. 

The UnicodeEncodeError occurs when running the scripts from cron. Agate is trying to encode the ellipsis character as ASCII. I presume this is because something in the cron environment is telling it to use ASCII (similar to https://github.com/wireservice/agate/issues/677).

I have tried everything I can think of to fix the server's locale so that LANG and LC_ALL are `en-US.UTF-8`. Everything *looks* fine. But when the cron job runs, the error persists.

The only way I have been able to overcome it is by specifying an encoding on the devnull open():

```
if sys.stdout.isatty():
    output_device = sys.stdout
else:
    output_device = open(os.devnull, 'w', encoding='utf8')
```

Anyway, long story short, it'd be great if there was some way to specify an encoding in `print_table()`, `to_csv()`, etc, which I believe would also resolve #624 and #677. 

I also want to again suggest that you change `make_number_formatter()` to remove the hard-coded ellipsis character, and instead replace it with the `ellipsis_chars` variable set in `config.py`. Maybe it would look like something like this:

```
def make_number_formatter(decimal_places, add_ellipsis=False):
    ellipsis_chars = config.get_option('ellipsis_chars')
    fraction = u'0' * decimal_places
    ellipsis = ellipsis_chars if add_ellipsis else u''
    return u''.join([u'#,##0.', fraction, ellipsis, u';-#,##0.', fraction, ellipsis])
```

**Comment by jpmckinney** [comment]
Are you still using Python 2.7?

**Comment by Kirkman** [comment]
Yes. Hoping to move to Python 3 soon.

**Comment by jpmckinney** [comment]
Okay, closing since Python 2 is EOL so no longer supported.
